File: 1235.maximum-profit-in-job-scheduling.cpp
/*
 * @lc app=leetcode id=1235 lang=cpp
 *
 * [1235] Maximum Profit in Job Scheduling
 */

// @lc code=start
#include <vector>
#include <algorithm>
#include <iostream>

class Solution {
public:
    int jobScheduling(std::vector<int>& startTime, std::vector<int>& endTime, std::vector<int>& profit) {
        std::vector<std::vector<int>> jobs;

        for (int i = 0; i < startTime.size(); ++i) {
            jobs.push_back({startTime[i], endTime[i], profit[i]});
        }

        std::sort(jobs.begin(), jobs.end(), [](const auto& a, const auto& b) {
            return a[0] < b[0];
        });

        int n = jobs.size();
        std::vector<int> result(n, 0);

        for (int i = n - 1; i >= 0; --i) {
            int currentProfit = jobs[i][2];
            auto it = std::lower_bound(jobs.begin(), jobs.end(), jobs[i][1], [](const auto& a, int val) {
                return a[0] < val;
            });

            int index = it - jobs.begin();

            if (index < n) {
                currentProfit += result[index];
            }

            if (i < n - 1) {
                result[i] = result[i + 1];
            }

            result[i] = std::max(result[i], currentProfit);
        }

        return result[0];
    }
};

// @lc code=end

File: 1235.maximum-profit-in-job-scheduling.py
#
# @lc app=leetcode id=1235 lang=python3
#
# [1235] Maximum Profit in Job Scheduling
#

# @lc code=start
import bisect
from operator import itemgetter
from typing import List


class Solution:
    '''
    Well, since profit can vary independently of start time and end time, we should think about DP approach (structured brute-force)
    Firstly, group [startTime, endTime, profit] together: jobs
    Then sort according to startTime
    dp[jobs[i]] = max profit possible given jobs[i:]
    we want max(dp[:])    
    '''

    def jobScheduling(self, startTime: List[int], endTime: List[int], profit: List[int]) -> int:
        jobs = []
        for s, e, p in zip(startTime, endTime, profit):
            jobs.append([s, e, p])
        jobs.sort(key = lambda x: x[0])
        n = len(jobs)
        result = [0] * n
        for i in range(n - 1, -1, -1):
            currentProfit = jobs[i][2]
            index = bisect.bisect_left(jobs, jobs[i][1], key = itemgetter(0))
            if index < n: currentProfit += result[index]
            if i < n - 1: result[i] = result[i + 1]
            result[i] = max(result[i], currentProfit)
        return result[0]
# @lc code=end

